File: socai/agent/loop.py
"""Minimal agent loop.

This module is intentionally browser-agnostic. CDP/browser tools should be
registered as ``Tool`` instances by a higher layer.
"""

from __future__ import annotations

import asyncio
import hashlib
import json
import time
from pathlib import Path

from .backends import Backend, create_backend
from .run_logging import RunDebugLogger, current_traceback, make_run_dir
from .run_state import RunState
from .tool import Tool, ToolContext


_BASE_SYSTEM_PROMPT = """\
You are a computer-use agent. Use the provided tools when they help complete
the user's task. Think briefly, take one or more useful actions, verify results
from tool output, and finish with a concise report when the task is complete.

Rules:
- Prefer high-level task/site tools over low-level manual actions when both exist.
- Do not invent observations. Use tool results as evidence.
- If a tool fails, explain the failure and choose a smaller recovery step.
- When enough evidence has been collected, stop calling tools and answer.
"""


def _build_system_prompt(tools: list[Tool], extra_instructions: str = "") -> str:
    parts = [_BASE_SYSTEM_PROMPT]
    if tools:
        parts.append(
            "Available tool names: "
            + ", ".join(f"`{tool.name}`" for tool in tools)
            + ". Tool schemas are provided separately."
        )
    if extra_instructions.strip():
        parts.append("Additional instructions:\n\n" + extra_instructions.strip())
    return "\n\n".join(parts)


def _tool_call_signature(tool_name: str, tool_input: dict) -> str:
    try:
        payload = json.dumps(tool_input or {}, sort_keys=True, ensure_ascii=False)
    except Exception:
        payload = str(tool_input)
    return hashlib.md5(f"{tool_name}::{payload}".encode("utf-8")).hexdigest()[:12]


def _text_summary(content_blocks: list[dict], max_len: int = 500) -> str:
    parts: list[str] = []
    for block in content_blocks:
        if block.get("type") == "text":
            parts.append(str(block.get("text", ""))[:max_len])
        elif block.get("type") == "image":
            parts.append("[image]")
    return " | ".join(parts)[:max_len]


def _compact_memory_entries(entries: list[str], max_chars: int) -> str:
    if not entries or max_chars <= 0:
        return ""
    selected: list[str] = []
    total = 0
    for entry in reversed(entries):
        entry = str(entry or "").strip()
        if not entry:
            continue
        projected = total + len(entry) + 1
        if selected and projected > max_chars:
            break
        selected.append(entry[:max_chars] if not selected and len(entry) > max_chars else entry)
        total = min(projected, max_chars)
        if total >= max_chars:
            break
    return "\n".join(reversed(selected))[:max_chars]


def _is_tool_result_message(message: dict) -> bool:
    content = message.get("content")
    if isinstance(content, str):
        return content.lstrip().startswith("[Tool result for ")
    if not isinstance(content, list):
        return False
    return any(
        isinstance(block, dict) and block.get("type") in {"tool_result", "function_call_output"}
        for block in content
    )


def _prepare_messages_for_context(
    messages: list[dict],
    run_state: RunState | None,
    memory_entries: list[str],
    *,
    keep_recent_messages: int = 12,
    memory_max_chars: int = 6000,
) -> list[dict]:
    if len(messages) <= keep_recent_messages + 2:
        return messages

    sections: list[str] = []
    if run_state is not None:
        state_block = run_state.context_block(max_chars=max(1200, memory_max_chars // 2))
        if state_block:
            sections.append("Structured run state from earlier turns:\n\n" + state_block)
    memory = _compact_memory_entries(memory_entries, memory_max_chars)
    if memory:
        sections.append("Condensed event memory from earlier turns:\n\n" + memory)
    if not sections:
        return messages

    recent = list(messages[-keep_recent_messages:])
    while recent and _is_tool_result_message(recent[0]):
        recent.pop(0)
    return [messages[0], {"role": "user", "content": "\n\n".join(sections)}, *recent]


async def _execute_tool(tool: Tool, params: dict, ctx: ToolContext) -> list[dict]:
    result = await tool.execute(params, ctx)
    if isinstance(result, str):
        return [{"type": "text", "text": result}]
    if isinstance(result, list):
        return result
    return [{"type": "text", "text": str(result)}]


async def run_agent(
    task: str,
    *,
    backend: Backend | None = None,
    tools: list[Tool] | None = None,
    run_dir: str | Path | None = None,
    max_turns: int = 30,
    model: str | None = None,
    extra_instructions: str = "",
    log_callback=None,
) -> dict:
    """Run the generic agent loop.

    Args:
        task: User task.
        backend: Optional prebuilt backend. Tests and app hosts can inject one.
        tools: Concrete tools available to the agent. Browser/CDP tools are
            intentionally supplied by the caller.
        run_dir: Output directory for report, logs, and run state.
        max_turns: Maximum model turns.
        model: Model id used when ``backend`` is not provided.
        extra_instructions: Additional system prompt text.
        log_callback: Optional ``callable(event, detail)`` for host logging.
    """

    if run_dir is None:
        run_dir = make_run_dir(task)
    run_path = Path(run_dir)
    run_path.mkdir(parents=True, exist_ok=True)

    available_tools = list(tools or [])
    tool_map = {tool.name: tool for tool in available_tools}
    backend = backend or create_backend(model)
    selected_model = str(model or getattr(backend, "model", "") or "")
    run_state = RunState(run_dir=run_path, task=task, model=selected_model)
    ctx = ToolContext(run_dir=run_path, run_state=run_state)
    debug_log = RunDebugLogger(run_path)

    def log(event: str, detail: str = "") -> None:
        if log_callback:
            log_callback(event, detail)

    messages: list[dict] = [{"role": "user", "content": task}]
    system_prompt = _build_system_prompt(available_tools, extra_instructions)
    api_tools = [tool.to_api_schema() for tool in available_tools]
    reasoning_log_path = debug_log.reasoning_log_path
    task_start = time.time()
    final_text = ""
    context_memory: list[str] = []
    tool_call_signatures: dict[str, list[int]] = {}
    turn = 0
    completed = False

    debug_log.event("task_start", task=task, model=selected_model, tools=[tool.name for tool in available_tools])
    log("start", task)

    while turn < max_turns:
        turn += 1
        turn_start = time.time()
        log("turn", f"{turn}/{max_turns}")

        request_messages = _prepare_messages_for_context(messages, run_state, context_memory)
        try:
            response = backend.create_message(
                system=system_prompt,
                messages=request_messages,
                tools=api_tools,
                max_tokens=8192,
            )
        except Exception as exc:  # noqa: BLE001 - backend boundary
            final_text = f"API error: {exc}"
            debug_log.api_error(turn=turn, error=str(exc))
            break

        messages.append({"role": "assistant", "content": backend.format_assistant_content(response)})
        visible_texts = [text for text in response.text_blocks if not text.startswith("[Thinking] ")]
        if visible_texts:
            final_text = "\n".join(visible_texts)

        run_state.note_assistant_turn(
            turn=turn,
            text="\n".join(visible_texts or response.text_blocks),
            tool_calls=[{"name": call.name, "input": call.input} for call in response.tool_calls],
        )
        debug_log.event(
            "llm_response",
            turn=turn,
            stop_reason=response.stop_reason,
            text="\n".join(visible_texts),
            tool_calls=[{"name": call.name, "input": call.input} for call in response.tool_calls],
            usage={"input_tokens": response.input_tokens, "output_tokens": response.output_tokens},
        )

        if not response.tool_calls:
            completed = True
            break

        all_results: list[list[dict]] = []
        for call_index, tool_call in enumerate(response.tool_calls, start=1):
            tool_name = tool_call.name
            tool_input = tool_call.input
            ctx.turn = turn
            ctx.active_tool_name = tool_name
            signature = _tool_call_signature(tool_name, tool_input)
            history = tool_call_signatures.setdefault(signature, [])
            history.append(turn)

            run_state.note_tool_call(turn=turn, tool_name=tool_name, tool_input=tool_input)
            debug_log.event(
                "tool_call_start",
                turn=turn,
                sequence=call_index,
                tool=tool_name,
                input=tool_input,
                repeat_count=len(history),
            )
            tool = tool_map.get(tool_name)
            tool_start = time.time()
            tool_error = ""
            tool_traceback = ""
            if tool is None:
                tool_error = f"Unknown tool '{tool_name}'"
                result_content = [{"type": "text", "text": f"Error: {tool_error}"}]
            else:
                try:
                    result_content = await _execute_tool(tool, tool_input, ctx)
                except Exception as exc:  # noqa: BLE001 - tool boundary
                    tool_error = str(exc)
                    tool_traceback = current_traceback()
                    result_content = [{"type": "text", "text": f"Error executing {tool_name}: {exc}"}]

            duration_s = round(time.time() - tool_start, 2)
            summary = _text_summary(result_content, max_len=900)
            debug_log.tool_result(
                turn=turn,
                sequence=call_index,
                tool_name=tool_name,
                tool_input=tool_input,
                content=result_content,
                duration_s=duration_s,
                result_summary=summary,
                repeat_count=len(history),
                error=tool_error,
                traceback_text=tool_traceback,
            )
            run_state.note_tool_result(
                turn=turn,
                tool_name=tool_name,
                tool_input=tool_input,
                result_summary=summary,
                duration_s=duration_s,
            )
            context_memory.append(
                f"- turn {turn} {tool_name}({json.dumps(tool_input, ensure_ascii=False)[:160]}): {summary}"
            )
            context_memory = context_memory[-80:]
            all_results.append(result_content)
            ctx.active_tool_name = ""

        messages.append(backend.format_tool_results(response.tool_calls, all_results))
        await asyncio.sleep(0)
        debug_log.event("turn_end", turn=turn, duration_s=round(time.time() - turn_start, 2))

    if turn >= max_turns and not completed:
        log("turn", f"final-summary (max_turns={max_turns} reached)")
        messages.append(
            {
                "role": "user",
                "content": (
                    f"You have reached the maximum of {max_turns} tool-using turns. "
                    "Do not call any more tools. Based on the evidence already gathered, "
                    "produce the best possible final answer for the user now in the same "
                    "language as the original task. If information is incomplete, state "
                    "what is known, what is missing, and give your best-effort conclusion."
                ),
            }
        )
        request_messages = _prepare_messages_for_context(messages, run_state, context_memory)
        try:
            response = backend.create_message(
                system=system_prompt,
                messages=request_messages,
                tools=[],
                max_tokens=8192,
            )
            messages.append({"role": "assistant", "content": backend.format_assistant_content(response)})
            visible_texts = [text for text in response.text_blocks if not text.startswith("[Thinking] ")]
            if visible_texts:
                final_text = "\n".join(visible_texts)
            run_state.note_assistant_turn(
                turn=turn + 1,
                text="\n".join(visible_texts or response.text_blocks),
                tool_calls=[],
            )
            debug_log.event(
                "llm_response",
                turn=turn + 1,
                stop_reason=response.stop_reason,
                text="\n".join(visible_texts),
                tool_calls=[],
                usage={"input_tokens": response.input_tokens, "output_tokens": response.output_tokens},
                forced_summary=True,
            )
        except Exception as exc:  # noqa: BLE001 - backend boundary
            suffix = f"Reached max_turns ({max_turns}) and forced-summary call failed: {exc}"
            final_text = f"{final_text}\n\n{suffix}".strip() if final_text else suffix
            debug_log.api_error(turn=turn + 1, error=str(exc), forced_summary=True)

    total_duration = round(time.time() - task_start, 2)
    report_path = run_path / "report.md"
    report_path.write_text(final_text, encoding="utf-8")
    conversation_path = debug_log.write_conversation(system_prompt=system_prompt, messages=messages)
    summary = {
        "task": task,
        "model": selected_model,
        "turns": turn,
        "total_duration_s": total_duration,
        "run_dir": str(run_path),
        "run_state_dir": str(run_state.state_dir),
        "reasoning_log_file": "reasoning_log.jsonl",
        "conversation_file": "conversation.json",
        "report_file": "report.md",
        "tool_results_dir": "tool_results",
    }
    debug_log.write_agent_summary(summary)
    debug_log.event("task_end", turn=turn, total_duration_s=total_duration)
    log("done", str(run_path))

    return {
        "result": final_text,
        "turns": turn,
        "run_dir": str(run_path),
        "run_state_dir": str(run_state.state_dir),
        "reasoning_log": str(reasoning_log_path),
        "conversation": str(conversation_path),
        "total_duration_s": total_duration,
    }

File: core/src/agent/loop.rs
//! Agent loop — the heart of the agent runtime.
//!
//! ```text
//!   while turn < max_turns:
//!     response = backend.send(system, messages, tool_schemas)
//!     append assistant
//!     if no tool calls: break
//!     for tc in tool_calls:
//!       result = dispatcher.call(tc)
//!       append tool_result
//! ```
//!
//! Cross-cutting concerns are split out:
//! - `signature.rs` — md5 fingerprint for repeated-call detection
//! - `memory.rs`    — windowing the message history once it's long
//! - `report.rs`    — final report enrichment with artifact links
//! - `compaction.rs` — truncating tool_result bodies for the history budget
//! - `run_state.rs` / `run_logging.rs` — persisting events + artifacts

use std::collections::BTreeMap;
use std::path::{Path, PathBuf};
use std::sync::Arc;
use std::time::Instant;

use serde_json::{json, Value};
use tokio::sync::broadcast;
use tracing::{debug, info, warn};

use crate::agent::compaction::{compress_text_maybe_json, TOOL_RESULT_TEXT_MAX_CHARS};
use crate::agent::llm::{
    Backend, Block, LLMResponse, Message, StopReason, ToolCall, ToolResultContent, ToolSchema,
};
use crate::agent::memory::prepare_messages_for_context;
use crate::agent::report::report_with_artifacts;
use crate::agent::run_logging::{make_run_dir, RunDebugLogger};
use crate::agent::run_state::RunState;
use crate::agent::signature::tool_call_signature;
use crate::agent::system_prompt::build_system_prompt;
use crate::agent::tool::{SharedTool, ToolContext, ToolResult, ToolResultBlock};

/// Events streamed to subscribers while the agent is running.
#[derive(Debug, Clone)]
pub enum AgentEvent {
    Started {
        run_id: String,
        task: String,
        model: String,
    },
    Turn {
        turn: u32,
    },
    AssistantText {
        turn: u32,
        text: String,
    },
    Reasoning {
        turn: u32,
        text: String,
    },
    ToolCall {
        turn: u32,
        name: String,
        input: Value,
        repeat_count: u32,
    },
    ToolResult {
        turn: u32,
        name: String,
        summary: String,
        duration_ms: u64,
        error: Option<String>,
    },
    ApiError {
        turn: u32,
        message: String,
    },
    Done {
        run_id: String,
        turns: u32,
        final_text: String,
    },
}

#[derive(Debug, Clone)]
pub struct AgentOptions {
    pub max_turns: u32,
    pub max_tokens: u32,
    pub extra_instructions: String,
    pub run_dir: Option<PathBuf>,
    /// Site names to pre-enable in ToolContext (gates `defer_until_site` tools).
    pub enabled_sites: Vec<String>,
    /// Recent-message window kept verbatim when history is condensed.
    pub keep_recent_messages: usize,
    /// Memory budget for the condensed context_block.
    pub memory_max_chars: usize,
}

impl Default for AgentOptions {
    fn default() -> Self {
        Self {
            max_turns: 30,
            max_tokens: 8192,
            extra_instructions: String::new(),
            run_dir: None,
            enabled_sites: Vec::new(),
            keep_recent_messages: 12,
            memory_max_chars: 6000,
        }
    }
}

#[derive(Debug, Clone)]
pub struct AgentOutcome {
    pub run_id: String,
    pub run_dir: PathBuf,
    pub turns: u32,
    pub final_text: String,
    pub total_input_tokens: u64,
    pub total_output_tokens: u64,
}

pub async fn run_agent(
    task: &str,
    backend: Arc<dyn Backend>,
    tools: Vec<SharedTool>,
    options: AgentOptions,
) -> anyhow::Result<AgentOutcome> {
    let (tx, _rx) = broadcast::channel(256);
    run_agent_with_events(task, backend, tools, options, tx).await
}

pub async fn run_agent_with_events(
    task: &str,
    backend: Arc<dyn Backend>,
    tools: Vec<SharedTool>,
    options: AgentOptions,
    events: broadcast::Sender<AgentEvent>,
) -> anyhow::Result<AgentOutcome> {
    let run_id = new_run_id();
    let run_dir = options.run_dir.unwrap_or_else(|| make_run_dir(task));
    ensure_dir(&run_dir)?;
    let run_state = Arc::new(RunState::new(&run_dir, task, backend.model())?);
    let debug_log = RunDebugLogger::new(&run_dir);

    let mut ctx = ToolContext::new(&run_id, &run_dir).with_run_state(Arc::clone(&run_state));
    for site in &options.enabled_sites {
        ctx.enable_site(site.clone());
    }

    let mut messages: Vec<Message> = vec![Message::user(task.to_string())];

    debug_log.event(
        "task_start",
        json!({
            "task": task,
            "model": backend.model(),
            "tools": tools.iter().map(|t| t.name()).collect::<Vec<_>>(),
        }),
    );
    emit(
        &events,
        AgentEvent::Started {
            run_id: run_id.clone(),
            task: task.to_string(),
            model: backend.label(),
        },
    );

    let mut turn = 0u32;
    let mut final_text = String::new();
    let mut total_input_tokens = 0u64;
    let mut total_output_tokens = 0u64;
    let mut context_memory: Vec<String> = Vec::new();
    let mut tool_call_history: BTreeMap<String, Vec<u32>> = BTreeMap::new();
    let mut completed = false;
    let mut last_system: String = build_system_prompt(&[], &options.extra_instructions);

    while turn < options.max_turns {
        turn += 1;
        ctx.turn = turn;
        emit(&events, AgentEvent::Turn { turn });
        debug!(turn, "agent turn start");

        let schemas = tool_schemas(&tools, &ctx);
        let tool_names: Vec<&str> = schemas.iter().map(|s| s.name.as_str()).collect();
        let system = build_system_prompt(&tool_names, &options.extra_instructions);
        last_system = system.clone();
        let request_messages = prepare_messages_for_context(
            &messages,
            Some(&run_state),
            &context_memory,
            options.keep_recent_messages,
            options.memory_max_chars,
        );

        let response: LLMResponse = match backend
            .send(&system, &request_messages, &schemas, options.max_tokens)
            .await
        {
            Ok(r) => r,
            Err(e) => {
                let msg = format!("{e:#}");
                warn!(turn, error = %msg, "backend error");
                emit(
                    &events,
                    AgentEvent::ApiError {
                        turn,
                        message: msg.clone(),
                    },
                );
                debug_log.api_error(turn, &msg, false);
                final_text = format!("API error: {msg}");
                break;
            }
        };

        total_input_tokens += response.input_tokens;
        total_output_tokens += response.output_tokens;

        // Split text_blocks into visible vs "[Thinking] "-prefixed thinking.
        // Some hosts (Anthropic without extended-thinking enabled) ask the
        // model to prefix its reasoning so we can keep it out of final_text
        // while still emitting it on the event stream for UIs that want to
        // show it.
        let (visible_texts, thinking_texts) = split_thinking(&response.text_blocks);

        // Surface reasoning to subscribers — both the structured
        // reasoning_content (Kimi/Qwen) and the [Thinking]-prefixed text.
        if !response.reasoning_content.trim().is_empty() {
            emit(
                &events,
                AgentEvent::Reasoning {
                    turn,
                    text: response.reasoning_content.clone(),
                },
            );
        }
        if !thinking_texts.is_empty() {
            emit(
                &events,
                AgentEvent::Reasoning {
                    turn,
                    text: thinking_texts.join("\n"),
                },
            );
        }

        // Build the assistant block list manually instead of using
        // LLMResponse::to_assistant_blocks() so we can:
        // - drop [Thinking]-prefixed text from history
        // - truncate visible text to ASSISTANT_TEXT_MAX_CHARS, matching
        //   Python's format_assistant_content (320 chars)
        let assistant_blocks = build_assistant_blocks(&response, &visible_texts);
        messages.push(Message::assistant_blocks(assistant_blocks));

        for text in &visible_texts {
            emit(
                &events,
                AgentEvent::AssistantText {
                    turn,
                    text: text.clone(),
                },
            );
            final_text = text.clone();
        }

        let tool_call_summary: Vec<Value> = response
            .tool_calls
            .iter()
            .map(|tc| json!({"name": tc.name, "input": tc.input}))
            .collect();
        run_state.note_assistant_turn(turn, &visible_texts.join("\n"), &tool_call_summary);
        debug_log.event(
            "llm_response",
            json!({
                "turn": turn,
                "stop_reason": stop_reason_str(response.stop_reason),
                "text": visible_texts.join("\n"),
                "tool_calls": tool_call_summary,
                "usage": {
                    "input_tokens": response.input_tokens,
                    "output_tokens": response.output_tokens,
                },
            }),
        );

        if response.tool_calls.is_empty() {
            completed = true;
            break;
        }

        let mut tool_result_blocks: Vec<Block> = Vec::new();
        for (idx, tc) in response.tool_calls.iter().enumerate() {
            let ToolCall { id, name, input } = tc;
            ctx.active_tool_name = name.clone();

            let sig = tool_call_signature(name, input);
            let history = tool_call_history.entry(sig).or_default();
            history.push(turn);
            let repeat_count = history.len() as u32;

            emit(
                &events,
                AgentEvent::ToolCall {
                    turn,
                    name: name.clone(),
                    input: input.clone(),
                    repeat_count,
                },
            );
            run_state.note_tool_call(turn, name, input);
            debug_log.event(
                "tool_call_start",
                json!({
                    "turn": turn,
                    "sequence": idx + 1,
                    "tool": name,
                    "input": input,
                    "repeat_count": repeat_count,
                }),
            );

            let started = Instant::now();
            let (result, error) = dispatch_tool(&tools, name, input, &ctx).await;
            let duration_ms = started.elapsed().as_millis() as u64;
            let duration_s = (duration_ms as f64) / 1000.0;

            let result_content = tool_result_to_content(&result);
            let flat = result.flat_text();
            let summary = truncate_summary(&flat, 240);
            emit(
                &events,
                AgentEvent::ToolResult {
                    turn,
                    name: name.clone(),
                    summary: summary.clone(),
                    duration_ms,
                    error: error.clone(),
                },
            );
            run_state.note_tool_result(turn, name, input, &summary, duration_s);
            debug_log.tool_result(
                turn,
                (idx + 1) as u32,
                name,
                input,
                &content_for_log(&result_content),
                duration_s,
                &summary,
                repeat_count,
                error.as_deref().unwrap_or(""),
            );

            tool_result_blocks.push(Block::ToolResult {
                tool_use_id: id.clone(),
                content: bound_content_for_history(&result_content),
            });

            context_memory.push(format!(
                "- turn {turn} {name}({}): {summary}",
                truncate_summary(&serde_json::to_string(input).unwrap_or_default(), 160)
            ));
            if context_memory.len() > 80 {
                let overflow = context_memory.len() - 80;
                context_memory.drain(0..overflow);
            }
            ctx.active_tool_name.clear();
        }
        messages.push(Message::user_blocks(tool_result_blocks));
        debug_log.event("turn_end", json!({"turn": turn}));
    }

    if !completed && turn >= options.max_turns {
        info!(turn, "reached max_turns, forcing final summary");
        messages.push(Message::user(format!(
            "You have reached the maximum of {} tool-using turns. Do not call any \
             more tools. Based on the evidence already gathered, produce the best \
             possible final answer for the user now in the same language as the \
             original task. If information is incomplete, state what is known, \
             what is missing, and give your best-effort conclusion.",
            options.max_turns
        )));
        let request_messages = prepare_messages_for_context(
            &messages,
            Some(&run_state),
            &context_memory,
            options.keep_recent_messages,
            options.memory_max_chars,
        );
        match backend
            .send(&last_system, &request_messages, &[], options.max_tokens)
            .await
        {
            Ok(response) => {
                total_input_tokens += response.input_tokens;
                total_output_tokens += response.output_tokens;
                let (visible_texts, _) = split_thinking(&response.text_blocks);
                for text in &visible_texts {
                    emit(
                        &events,
                        AgentEvent::AssistantText {
                            turn: turn + 1,
                            text: text.clone(),
                        },
                    );
                    final_text = text.clone();
                }
                debug_log.event(
                    "llm_response",
                    json!({
                        "turn": turn + 1,
                        "forced_summary": true,
                        "text": final_text,
                        "usage": {
                            "input_tokens": response.input_tokens,
                            "output_tokens": response.output_tokens,
                        },
                    }),
                );
            }
            Err(e) => {
                let msg = format!("{e:#}");
                warn!(turn = turn + 1, error = %msg, "forced summary error");
                emit(
                    &events,
                    AgentEvent::ApiError {
                        turn: turn + 1,
                        message: msg.clone(),
                    },
                );
                debug_log.api_error(turn + 1, &msg, true);
            }
        }
    }

    emit(
        &events,
        AgentEvent::Done {
            run_id: run_id.clone(),
            turns: turn,
            final_text: final_text.clone(),
        },
    );

    let enriched_report = report_with_artifacts(&final_text, Some(&run_state));
    let _ = std::fs::write(run_dir.join("report.md"), &enriched_report);

    let serialized_messages = serde_json::to_value(&messages).unwrap_or(Value::Null);
    debug_log.write_conversation(&last_system, &serialized_messages);

    let summary = json!({
        "task": task,
        "model": backend.label(),
        "turns": turn,
        "run_dir": run_dir.to_string_lossy(),
        "input_tokens": total_input_tokens,
        "output_tokens": total_output_tokens,
        "reasoning_log_file": "reasoning_log.jsonl",
        "conversation_file": "conversation.json",
        "report_file": "report.md",
        "tool_results_dir": "tool_results",
    });
    debug_log.write_agent_summary(&summary);
    debug_log.event("task_end", json!({"turn": turn, "completed": completed}));

    Ok(AgentOutcome {
        run_id,
        run_dir,
        turns: turn,
        final_text,
        total_input_tokens,
        total_output_tokens,
    })
}

// ---------- small private helpers (not core logic, kept here for locality) ----------

fn new_run_id() -> String {
    use chrono::Utc;
    use std::time::{SystemTime, UNIX_EPOCH};
    let suffix = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.subsec_nanos() % 1_000_000)
        .unwrap_or(0);
    format!("{}-{:06}", Utc::now().format("%Y%m%d-%H%M%S"), suffix)
}

fn ensure_dir(path: &Path) -> anyhow::Result<()> {
    std::fs::create_dir_all(path)?;
    Ok(())
}

fn tool_schemas(tools: &[SharedTool], ctx: &ToolContext) -> Vec<ToolSchema> {
    tools
        .iter()
        .filter(|t| t.is_available(ctx))
        .map(|t| ToolSchema {
            name: t.name().to_string(),
            description: t.description().to_string(),
            input_schema: t.input_schema(),
        })
        .collect()
}

fn find_tool<'a>(tools: &'a [SharedTool], name: &str) -> Option<&'a SharedTool> {
    tools.iter().find(|t| t.name() == name)
}

fn emit(events: &broadcast::Sender<AgentEvent>, event: AgentEvent) {
    let _ = events.send(event);
}

fn stop_reason_str(reason: StopReason) -> &'static str {
    match reason {
        StopReason::EndTurn => "end_turn",
        StopReason::ToolUse => "tool_use",
        StopReason::MaxTokens => "max_tokens",
        StopReason::Other => "other",
    }
}

async fn dispatch_tool(
    tools: &[SharedTool],
    name: &str,
    input: &Value,
    ctx: &ToolContext,
) -> (ToolResult, Option<String>) {
    match find_tool(tools, name) {
        Some(tool) if tool.is_available(ctx) => match tool.call(input.clone(), ctx).await {
            Ok(r) => (r, None),
            Err(e) => {
                let msg = format!("{e:#}");
                (
                    ToolResult::text(format!("Error executing {name}: {msg}")),
                    Some(msg),
                )
            }
        },
        Some(_) => {
            let msg = format!("Tool '{name}' is not currently available");
            (ToolResult::text(format!("Error: {msg}")), Some(msg))
        }
        None => {
            let msg = format!("Unknown tool '{name}'");
            (ToolResult::text(format!("Error: {msg}")), Some(msg))
        }
    }
}

fn tool_result_to_content(result: &ToolResult) -> Vec<ToolResultContent> {
    result
        .blocks
        .iter()
        .map(|b| match b {
            ToolResultBlock::Text { text } => ToolResultContent::Text { text: text.clone() },
            ToolResultBlock::Image { data, media_type } => ToolResultContent::Image {
                data: data.clone(),
                media_type: media_type.clone(),
            },
        })
        .collect()
}

/// Squash a tool_result for the chat history. Mirrors Python's
/// `_summarize_result_blocks_for_history`:
/// - text blocks → compressed JSON-aware truncation
/// - image blocks → text placeholder. If a preceding text block contained
///   "Screenshot saved to <path>", the placeholder names that path so the
///   model can still cite it in the final report.
///
/// Returns a single Text block (or `(empty result)` when nothing usable
/// remained). The raw bodies still hit disk via tool_results/*.json, so
/// nothing is lost — we just keep the chat-history budget bounded.
fn bound_content_for_history(content: &[ToolResultContent]) -> Vec<ToolResultContent> {
    let mut screenshot_path: Option<String> = None;
    let mut parts: Vec<String> = Vec::new();
    for block in content {
        match block {
            ToolResultContent::Text { text } => {
                if screenshot_path.is_none() {
                    screenshot_path = extract_screenshot_hint(text);
                }
                let compressed = compress_text_maybe_json(text, TOOL_RESULT_TEXT_MAX_CHARS);
                if !compressed.trim().is_empty() {
                    parts.push(compressed);
                }
            }
            ToolResultContent::Image { .. } => {
                parts.push(match &screenshot_path {
                    Some(path) => format!("[Image omitted from history. Screenshot file: {path}.]"),
                    None => "[Image omitted from history.]".to_string(),
                });
            }
        }
    }
    let mut combined = parts.join("\n\n").trim().to_string();
    if combined.chars().count() > TOOL_RESULT_TEXT_MAX_CHARS {
        combined = compress_text_maybe_json(&combined, TOOL_RESULT_TEXT_MAX_CHARS);
    }
    if combined.is_empty() {
        combined = "(empty result)".to_string();
    }
    vec![ToolResultContent::Text { text: combined }]
}

/// `"Screenshot saved to /tmp/x.png"` → `Some("/tmp/x.png")`. Mirrors
/// `_screenshot_hint_from_text` from Python.
fn extract_screenshot_hint(text: &str) -> Option<String> {
    let marker = "Screenshot saved to ";
    let idx = text.find(marker)?;
    let after = &text[idx + marker.len()..];
    let end = after
        .find(|c: char| c.is_whitespace())
        .unwrap_or(after.len());
    let path = after[..end].trim();
    if path.is_empty() {
        None
    } else {
        Some(path.to_string())
    }
}

/// Render tool-result content as a JSON value suitable for the
/// `reasoning_log.jsonl` debug stream (images are mentioned but bodies
/// omitted — same convention as `json_safe_for_log`).
fn content_for_log(content: &[ToolResultContent]) -> Value {
    let array: Vec<Value> = content
        .iter()
        .map(|c| match c {
            ToolResultContent::Text { text } => json!({"type": "text", "text": text}),
            ToolResultContent::Image { media_type, .. } => json!({
                "type": "image",
                "media_type": media_type,
            }),
        })
        .collect();
    Value::Array(array)
}

fn truncate_summary(text: &str, max_chars: usize) -> String {
    if text.chars().count() <= max_chars {
        return text.to_string();
    }
    let mut s: String = text.chars().take(max_chars).collect();
    s.push('…');
    s
}

/// Split assistant text blocks into (visible, thinking) by the `[Thinking] `
/// prefix. Whitespace-only blocks are dropped from both buckets. Mirrors the
/// Python `loop.py` slicing.
fn split_thinking(text_blocks: &[String]) -> (Vec<String>, Vec<String>) {
    const PREFIX: &str = "[Thinking] ";
    let mut visible: Vec<String> = Vec::new();
    let mut thinking: Vec<String> = Vec::new();
    for block in text_blocks {
        let trimmed = block.trim();
        if trimmed.is_empty() {
            continue;
        }
        if let Some(rest) = trimmed.strip_prefix(PREFIX) {
            thinking.push(rest.trim().to_string());
        } else {
            visible.push(trimmed.to_string());
        }
    }
    (visible, thinking)
}

/// Build the assistant message blocks for history. Truncates visible text
/// to `ASSISTANT_TEXT_MAX_CHARS` (320 — same as Python) to keep history
/// bounded over many turns. Drops `[Thinking]`-prefixed text since that's
/// already surfaced as a Reasoning event.
fn build_assistant_blocks(response: &LLMResponse, visible_texts: &[String]) -> Vec<Block> {
    use crate::agent::compaction::{truncate, ASSISTANT_TEXT_MAX_CHARS};
    let mut blocks: Vec<Block> = Vec::new();
    // Preserve reasoning_content alongside tool_calls so providers that
    // require it round-tripped (Kimi/Qwen) get it on the next turn.
    if !response.reasoning_content.trim().is_empty() && !response.tool_calls.is_empty() {
        blocks.push(Block::ReasoningContent {
            text: response.reasoning_content.clone(),
        });
    }
    for text in visible_texts {
        let bounded = truncate(text, ASSISTANT_TEXT_MAX_CHARS);
        if !bounded.is_empty() {
            blocks.push(Block::Text { text: bounded });
        }
    }
    for tc in &response.tool_calls {
        blocks.push(Block::ToolUse {
            id: tc.id.clone(),
            name: tc.name.clone(),
            input: tc.input.clone(),
        });
    }
    blocks
}
